x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java

```diff
@@ -7,55 +7,212 @@
-            return queryExec.canPushSorts();
+    }
+
+    private static Pushable evaluatePushable(TopNExec topNExec, Predicate<FieldAttribute> hasIdenticalDelegate) {
+        PhysicalPlan child = topNExec.child();
+        if (child instanceof EsQueryExec queryExec
+            && queryExec.canPushSorts()
+            && canPushDownOrders(topNExec.order(), hasIdenticalDelegate)) {
+            // With the simplest case of `FROM index | SORT ...` we only allow pushing down if the sort is on a field
+            return new PushableQueryExec(queryExec);
+        }
+        if (child instanceof ExchangeExec exchangeExec
+            && exchangeExec.child() instanceof EsQueryExec queryExec
+            && queryExec.canPushSorts()
+            && canPushDownOrders(topNExec.order(), hasIdenticalDelegate)) {
+            // When we have an exchange between the FROM and the SORT, we also only allow pushing down if the sort is on a field
+            return new PushableExchangeExec(exchangeExec, queryExec);
+        }
+        if (child instanceof EvalExec evalExec && evalExec.child() instanceof EsQueryExec queryExec && queryExec.canPushSorts()) {
+            // When we have an EVAL between the FROM and the SORT, we consider pushing down if the sort is on a field and/or
+            // a distance function defined in the EVAL. We also move the EVAL to after the SORT.
+            List<Order> orders = topNExec.order();
+            List<Alias> fields = evalExec.fields();
+            LinkedHashMap<NameId, StDistance> distances = new LinkedHashMap<>();
+            AttributeMap.Builder<Attribute> aliasReplacedByBuilder = AttributeMap.builder();
+            fields.forEach(alias -> {
+                // TODO: can we support CARTESIAN also?
+                if (alias.child() instanceof StDistance distance && distance.crsType() == BinarySpatialFunction.SpatialCrsType.GEO) {
+                    distances.put(alias.id(), distance);
+                } else if (alias.child() instanceof Attribute attr) {
+                    aliasReplacedByBuilder.put(alias.toAttribute(), attr.toAttribute());
+                }
+            });
+            AttributeMap<Attribute> aliasReplacedBy = aliasReplacedByBuilder.build();
+
+            List<EsQueryExec.Sort> pushableSorts = new ArrayList<>();
+            for (Order order : orders) {
+                if (LucenePushDownUtils.isPushableFieldAttribute(order.child(), hasIdenticalDelegate)) {
+                    pushableSorts.add(
+                        new EsQueryExec.FieldSort(
+                            ((FieldAttribute) order.child()).exactAttribute(),
+                            order.direction(),
+                            order.nullsPosition()
+                        )
+                    );
+                } else if (order.child() instanceof ReferenceAttribute referenceAttribute) {
+                    if (distances.containsKey(referenceAttribute.id())) {
```

Comment:
nit: This doesn't account for multiple aliasing, as in `EVAL dist = st_distance(x, "POINT(0 0)"::geo_point), dist2 = dist, dist3 = dist2 | SORT dist3 DESC`.

I think the correct way to do this is to 1. resolve the ref attribute using `aliasReplacedBy` (AttributeMap should take multiple indirections into account), and then 2. check if what we resolved to is either a pushable field attr or one of the pushable distances.

This would also do away with the code duplication in lines 168 and 189.

Reply:
Did a two line fix to cover this case (and expanded and existing test to test it). Did not reduce code duplication, because we cannot resolve the alias without first checking the type, and having that type check is kind-of the code duplication part. Perhaps some cleanup is possible, but not a complete de-dupliciation.